### PR TITLE
Refactor IS examples using pool-based QI

### PR DESCRIPTION
**Pull request by shazqadeer**
This PR does the following:

- It replaces triggers with pool-based QI in several IS examples.  

- It fixes a potential problem where the translation of the expression in an assert command was being used with the subsumption feature in two places in the generated VC, creating undesirable aliasing.  This problem was fixed by translating the expression twice for use in the two different contexts.

- It adds a flag /trustInductiveSequentialization. The default value is false.  If the user sets it to true, the IS checks are omitted.

- It adds resolution and type checking of attributes in assignments.

### CODE REVIEW

**Comment by shazqadeer** [comment]
@bkragl :  I have updated many of the IS examples to use pool-based QI.  The two versions of Chang-Roberts and Paxos are left.  These are more complicated and I would love to get your help with them.

**Comment by shazqadeer** [comment]
Two issues were discovered while working on this PR:
- Pool-related attributes on assignments are not being propagated.
- In the Paxos proof, the pool annotations on A_Join' and A_Vote' had to be added on a separate (dummy) "assume true".  An earlier attempt where those annotations were put on an existing assert in the actions did not work.

File: Source/Concurrency/InductiveSequentialization.cs [hunk 1]
```diff
@@ -319,6 +319,10 @@ public static class InductiveSequentializationChecker
   {
     public static void AddCheckers(CivlTypeChecker civlTypeChecker)
     {
+      if (CommandLineOptions.Clo.TrustInductiveSequentialization)
```
Comment (bkragl) [inline Source/Concurrency/InductiveSequentialization.cs:322]: nit: Move the conditional to the call site of `AddCheckers`.

File: Test/civl/inductive-sequentialization/2PC.bpl [hunk 2]
```diff
@@ -77,20 +75,18 @@ INV4 ({:linear_in "pid"} pids:[int]bool)
 returns ({:pending_async "PARTICIPANT2"} PAs:[PA]int, {:choice} choice:PA)
 modifies ReqCH, VoteCH, DecCH, votes, decisions;
 {
-  assert Init(pids, ReqCH, VoteCH, DecCH, decisions);
-
-  havoc ReqCH, VoteCH, DecCH, votes, decisions;
+  var {:pool "INV1"} k: int;
 
+  assert Init(pids, ReqCH, VoteCH, DecCH, decisions);
+  assume {:add_to_pool "INV1", k, k+1} {:add_to_pool "PARTICIPANT2", PARTICIPANT2(n)} 0 <= k && k <= n;
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/2PC.bpl:81]:
```suggestion
  assume
    {:add_to_pool "INV1", k, k+1}
    {:add_to_pool "PARTICIPANT2", PARTICIPANT2(n)}
    0 <= k && k <= n;
```

I find it more readable like this (and similar in other places). What do you think?

File: Test/civl/inductive-sequentialization/BroadcastConsensus.bpl [hunk 3]
```diff
@@ -93,23 +89,19 @@ INV_COLLECT_ELIM({:linear_in "broadcast"} pidsBroadcast:[pid]bool, {:linear_in "
...
-
-  assume
-  (exists k:int :: {trigger(k)} 0 <= k && k <= n && trigger(k+1) &&
-    // (forall i:pid :: 1 <= i && i <= k ==> decision[i] == max(CH)) &&
-    decision == (lambda i:pid :: if 1 <= i && i <= k then max(CH) else old(decision)[i]) &&
-    PAs == RemainingCollects(k) &&
-    choice == COLLECT(k+1) &&
-    (k < n ==> PAs[COLLECT(n)] > 0) // help for the prover
-  );
+  assume {:add_to_pool "INV_COLLECT", k} {:add_to_pool "INV_COLLECT", k+1} {:add_to_pool "Collect", COLLECT(n)} 0 <= k && k <= n;
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/BroadcastConsensus.bpl:101]: Can `{:add_to_pool "INV_COLLECT", k} {:add_to_pool "INV_COLLECT", k+1}` just be `{:add_to_pool "INV_COLLECT", k, k+1}`?

File: Test/civl/inductive-sequentialization/BroadcastConsensus.bpl [hunk 4]
```diff
@@ -147,20 +137,17 @@ INV_BROADCAST_ELIM({:linear_in "broadcast"} pidsBroadcast:[pid]bool, {:linear_in
...
-  assume
-  (exists k:int :: {trigger(k)} 0 <= k && k <= n && trigger(k+1) &&
-    CH == (lambda v:val :: value_card(v, value, 1, k)) &&
-    card(CH) == k &&
-    MultisetSubsetEq(MultisetEmpty, CH) &&
-    PAs == MapAdd(RemainingBroadcasts(k), AllCollects()) &&
-    choice == BROADCAST(k+1) &&
-    (k < n ==> PAs[BROADCAST(n)] > 0) // help for the prover
-  );
+  assume {:add_to_pool "INV_BROADCAST", k} {:add_to_pool "INV_BROADCAST", k+1} {:add_to_pool "Broadcast", BROADCAST(n)} 0 <= k && k <= n;
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/BroadcastConsensus.bpl:145]: `{:add_to_pool "INV_BROADCAST", k, k+1}`?

File: Test/civl/inductive-sequentialization/ChangRoberts.bpl [hunk 5]
```diff
@@ -84,26 +81,27 @@ INV2 ({:linear_in "pid"} pids:[int]bool)
...
-  );
+  assume {:add_to_pool "INV2", k, next(k), n+1} true;
+
+  choice := P(k);
+  if (*) {
+    assume
+      pid(k) &&
+      (forall i:int :: pid(i) && between(max(id),i,k) ==> terminated[i]) &&
+      (forall i:int :: pid(i) && !between(max(id),i,k) ==> !terminated[i]);
+      PAs := (lambda pa:PA :: if is#P(pa) && pid(pid#P(pa)) && !between(max(id), pid#P(pa), k) then 1 else 0);
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/ChangRoberts.bpl:97]: unindent

File: Test/civl/inductive-sequentialization/ChangRoberts.bpl [hunk 6]
```diff
@@ -84,26 +81,27 @@ INV2 ({:linear_in "pid"} pids:[int]bool)
...
+  } else {
+    assume
+      k == n + 1 &&
+      (forall i:int :: pid(i) ==> terminated[i]);
+      PAs := NoPAs();
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/ChangRoberts.bpl:102]: unindent

File: Test/civl/inductive-sequentialization/PingPong.bpl [hunk 7]
```diff
@@ -36,32 +34,34 @@ INV ({:linear_in "pid"} ping_pid:int, {:linear_in "pid"} pong_pid:int)
...
-      ||
-      (ping_channel == EmptyChannel()[c := 1] && pong_channel == EmptyChannel() &&
-       PAs == NoPAs()[PONG(c+1, pong_pid) := 1][PING(c, ping_pid) := 1] &&
-       choice == PING(c, ping_pid))
-      ||
-      (ping_channel == EmptyChannel() && pong_channel == EmptyChannel() &&
-       PAs == NoPAs())
-    )
-  );
+  assume {:add_to_pool "INV", c} {:add_to_pool "INV", c+1} 0 < c;
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/PingPong.bpl:44]: `{:add_to_pool "INV", c, c+1}`?

File: Test/civl/inductive-sequentialization/paxos/PaxosAbstractions.bpl [hunk 8]
```diff
@@ -69,8 +66,8 @@ modifies decision, pendingAsyncs;
   assert value#VoteInfo(t#Some(voteInfo[r])) == v;
 
   /**************************************************************************/
-  assert (forall n': Node, v': Value, p': Permission :: pendingAsyncs[A_Vote(r, n', v', p')] == 0);
-  assert triggerRound(r);
+  assert {:add_to_pool "Round", r}
+  (forall n': Node, v': Value, p': Permission :: pendingAsyncs[A_Vote(r, n', v', p')] == 0);
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/paxos/PaxosAbstractions.bpl:70]: indent

File: Test/civl/inductive-sequentialization/paxos/PaxosSeq.bpl [hunk 9]
```diff
@@ -76,45 +78,40 @@ modifies joinedNodes, voteInfo, decision, pendingAsyncs;
...
-    else
-    (
-      (
-        PAs == FirstCasePAs(k, numRounds) &&
-        choice == A_StartRound(k+1, k+1) &&
+
+  assume {:add_to_pool "NumRounds", numRounds} {:add_to_pool "Round", k, k+1, numRounds} 0 <= k && k <= numRounds;
+
+  if (k == numRounds) {
+      PAs := NoPAs();
```
Comment (bkragl) [inline Test/civl/inductive-sequentialization/paxos/PaxosSeq.bpl:85]: Two-space indentation in all if cases.

**Review by bkragl — COMMENTED** [review]
This is super cool! I *love* seeing all the existentially quantified assumes turned into individual assignments and the trigger functions gone.

**Comment by shazqadeer** [comment]
@bkragl :  I have incorporated all of your feedback.

**Pull request merged**